repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -82,21 +88,57 @@ class Session(interpreter: Interpreter, stateChangedCallback: SessionState => Un
   def execute(code: String): Int = {
     val statementId = newStatementId.getAndIncrement()
     _statements(statementId) = new Statement(statementId, StatementState.Waiting, null)
+
     Future {
-      _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+      def stmtState = _statements(statementId).state
+
+      if (stmtState != StatementState.Cancelled) {
+        lock.synchronized {
+          if (stmtState != StatementState.Cancelled) {
+            _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+            _sc.foreach(
+              _.setJobGroup(statementId.toString, s"Job group for statement $statementId"))
+          }
+        }
+      }
+
+      val executeResult = if (stmtState != StatementState.Cancelled) {
+        executeCode(statementId, code)
+      } else {
+        null
+      }
 
-      _statements(statementId) =
-        new Statement(statementId, StatementState.Available, executeCode(statementId, code))
+      if (stmtState != StatementState.Cancelled) {
+        lock.synchronized {
+          if (stmtState != StatementState.Cancelled) {
+            _statements(statementId) =
+              new Statement(statementId, StatementState.Available, executeResult)
+          }
+        }
+      }
     }
+
     statementId
   }
 
+  def cancel(statementId: Int): Unit = {
```

Comment:
How and when to catch this exception, in which thread? The job will be cancelled immediately or later on or never indefinitely, depends on the scheduler and pipeline. I'm inclining to not fix this unless user really need to distinguish `cancelling` or `cancelled`.